### PR TITLE
[Cash App Pay] Add Support for buying a gift card product type using Cash App Pay.

### LINKED ISSUE #73: [Cash App Pay] Add support for buying a Gift card product type
### Describe the bug
Currently, customers are unable to purchase gift card products using the Cash App Pay payment method. Customers should be able to buy gift card products using the Cash App Pay payment method.

### To reproduce
1. Create a gift card product.
2. Go to the front end of the web store.
3. Add the gift card product to the cart and proceed to the checkout page.
4. Notice that only the credit card payment gateway is listed as an available option.

**Screenshots**

### Expected behavior
Customers should be able to buy gift card products using the Cash App Pay payment method.

**Environment (please complete the following information):**
 - WordPress Version 6.4.3
 - WooCommerce Version: 8.5.2
 - WooCommerce Square Plugin Version: 4.5.0
 - Browser [e.g. chrome, safari] and Version
 - Any other plugins installed: N/A

### Additional details

<details><summary>System status</summary>
  
  ```
  <!--If applicable, paste the system status here. Please ensure you redact or remove any identifying information. -->
  ```
</details>

### LINKED ISSUE #74: Gift Card not getting created when credit card transaction type is "Authorization" and "Charge Virtual-Only Orders" enabled.
### Describe the bug
A gift card is not being created upon the purchase of a gift card product when the credit card transaction type is set to "Authorization," and "Charge Virtual-Only Orders" is enabled.

image

### To reproduce
1. Navigate to Square Credit Card payment method options (WooCommerce > Settings > Payments > Square).
2. Update the transaction type to "Authorization" and enable "Charge Virtual-Only Orders."
3. Save the settings.
4. Create a gift card product.
5. Visit the front end of the web store.
6. Add the gift card product to the cart and proceed to the checkout page to place an order.
7. Go to edit the order and observe that there is no order note related to the gift card created (Gift card with number: xxxxx created and activated).

**Screenshots**

### Expected behavior

**Environment (please complete the following information):**
 - WordPress Version: 6.4.3
 - WooCommerce Version: 8.5.2
 - WooCommerce Square Plugin Version: 4.5.0
 - Browser [e.g. chrome, safari] and Version
 - Any other plugins installed: N/A

### Additional details

<details><summary>System status</summary>
  
  ```
  <!--If applicable, paste the system status here. Please ensure you redact or remove any identifying information. -->
  ```
</details>

### PR DESCRIPTION
**Pull request by iamdharmesh**
### All Submissions:

* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
PR adds support for buying a gift card product type using Cash App Pay. PR also fixes #74 to ensure the gift card is created when the credit card transaction type is "Authorization," and "Charge Virtual-Only Orders" is enabled 

Closes #73 
Closes #74 

#### Test #73
1. Create a gift card product.
1. Go to the product page on the front-end of the web store, choose "Buy a new gift card," enter details, and add the product to the cart.
1. Go to the checkout and verify that Cash App Pay is available in the listed payment methods.
1. Place an order using the Cash App Pay gateway and verify that the order is placed successfully.
1. Verify that the gift card is created and activated.
1. Verify that an email is sent to the recipient.
1. Go to the gift card product page again, choose "Add value to an existing gift card," and add the product to the cart.
Place an order using the Cash App Pay gateway and verify that the order is placed successfully.

#### Test #74
Detailed issue reproduce steps are given in the issue description. follow that and verify that the issue no longer exists.

### CODE REVIEW

File: includes/Framework/PaymentGateway/Payment_Gateway.php [hunk 1]
```diff
@@ -1834,6 +1834,119 @@ protected function do_transaction( $order ) {
...
+		$response = $this->get_api()->load_gift_card( $gan, $order );
+
+		/** @var \Square\Models\GiftCardActivity $gift_card_activity */
+		$gift_card_activity = $response->get_data();
+
+		if ( ! $gift_card_activity instanceof \Square\Models\CreateGiftCardActivityResponse ) {
+			return false;
+		}
+
+		return false;
```
Comment (dkotter) [inline includes/Framework/PaymentGateway/Payment_Gateway.php]: Looks like this method always returns `false`. Is that correct or should it return `true` here?

File: includes/Framework/PaymentGateway/Payment_Gateway.php [hunk 2]
```diff
@@ -1834,6 +1834,119 @@ protected function do_transaction( $order ) {
...
+
+			$gift_card_line_item_id = $line_item->getUid();
+			$gift_card_amount       = Square_Helper::number_format(
+				Money_Utility::cents_to_float(
+					$line_item->getTotalMoney()->getAmount()
+				)
+			);
+
+			$this->update_order_meta( $order, 'gift_card_line_item_id', $gift_card_line_item_id );
+			$this->update_order_meta( $order, 'gift_card_balance', $gift_card_amount );
```
Comment (dkotter) [inline includes/Framework/PaymentGateway/Payment_Gateway.php]: Do we need sanitization on these values or does this already happen in `update_order_meta`?

File: includes/Framework/PaymentGateway/Payment_Gateway.php [hunk 3]
```diff
@@ -1834,6 +1834,119 @@ protected function do_transaction( $order ) {
...
+
+		/** @var \Square\Models\GiftCardActivity $gift_card_activity */
+		$gift_card_activity = $response->get_data();
+
+		if ( ! $gift_card_activity instanceof \Square\Models\CreateGiftCardActivityResponse ) {
+			return false;
+		}
+
+		$gan = $gift_card_activity->getGiftCardActivity()->getGiftCardGan();
+		$this->update_order_meta( $order, 'gift_card_number', $gan );
```
Comment (dkotter) [inline includes/Framework/PaymentGateway/Payment_Gateway.php]: Same question here on sanitization

File: includes/Framework/PaymentGateway/Payment_Gateway.php (hunk 1 above)
Comment (iamdharmesh) [reply includes/Framework/PaymentGateway/Payment_Gateway.php]: It should return `true`. I didn't notice this, as these functions have just moved from the child class. Thanks for catching this.

File: includes/Framework/PaymentGateway/Payment_Gateway.php (hunk 2 above)
Comment (iamdharmesh) [reply includes/Framework/PaymentGateway/Payment_Gateway.php]: Don't have sanitization in the `update_order_meta` function. Addressed in https://github.com/woocommerce/woocommerce-square/pull/79/commits/23283ab6e0f54c62c4a0a1e1aeb850c994479c43

File: includes/Framework/PaymentGateway/Payment_Gateway.php (hunk 3 above)
Comment (iamdharmesh) [reply includes/Framework/PaymentGateway/Payment_Gateway.php]: Addressed in https://github.com/woocommerce/woocommerce-square/pull/79/commits/23283ab6e0f54c62c4a0a1e1aeb850c994479c43

**Review by dkotter — APPROVED** [review]

**Comment by qasumitbagthariya** [comment]
### QA Update ✅ 

-------------
I have checked this issue in the `enhancement/73` branch and working as expected.

I tested the following on this branch:

- Closes #73 ✅ 
- Closes #74 ✅ 
- Cash app payment with Transaction Type >"Charge" ✅ 
- Cash app payment with Transaction Type >"Authoriatioin" ✅ 
- Refund ✅ 


**Testing Environment** 

<Details>

- WordPress: 6.4.3
- Theme: Storefront 4.5.4
- WooCommerce - 8.6.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>

https://github.com/woocommerce/woocommerce-square/assets/67687255/4d57a598-1e6e-473e-8e90-a8142291461a


https://github.com/woocommerce/woocommerce-square/assets/67687255/5086b376-1694-49da-94fd-fccee2dcd172




Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR

**Review by qasumitbagthariya — APPROVED** [review]

**Review by a-danae — APPROVED** [review]
Things look good to me 👍 thanks for your work here, @iamdharmesh !

✅ Cash App Pay is available under the available payment methods when adding a gift card to the cart.
✅ When purchasing a gift card with Cash App Pay, the gift card gets created - The related order has a note saying "Gift card with number: xxxxx created and activated", and the gift card is listed in Square.
✅  When purchasing a gift card with Cash App Pay, selecting "Add value to an existing gidt card" does add the purchased amount to the selected existing gift card.

When the card transaction type is set to "Authorization" and "Charge Virtual-Only Orders" is enabled:
✅ The gift card gets created - The related order has a note saying "Gift card with number: xxxxx created and activated", and the gift card is listed in Square.
✅  Selecting "Add value to an existing gidt card" does add the purchased amount to the selected existing gift card.

**Comment by qasumitbagthariya** [comment]
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.


**Testing Environment** 

<Details>

- WordPress: 6.4.3
- Theme: Twenty Twenty-Four 1.0
- WooCommerce - 8.7.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: smoke-testing

</Details>

Next Step- Ready to Merge 🚀

Status update:
@vikrampm1

**Pull request merged**